### ISSUE TITLE
One .tt file per subfolder

### ISSUE THREAD

**Issue by ClassyCircuit** [ReversePOCO/EntityFramework-Reverse-POCO-Code-First-Generator#389]
Hi,
Would it be possible for you to modify the tool, so when we generate entities, context & mapping in separate folders, then it would use only one, single .tt file? Otherwise, it is cumbersome to set it up, if there are many databases.

Right now my setup is the following:
![image](https://user-images.githubusercontent.com/6424248/37087629-87e78b00-2203-11e8-83b0-913a6d1daf57.png)

What I would like to have is the same subfolder structure, but having the .tt source in only one place.

If you don't have time to do it, I am willing to help out and make a pull request, but then I would need some pointers where to start in your code.

Thanks!

**Comment by delphinusc** [comment]
Can this be reopened as an enhancement?  There's only one set of Settings values that are really specific to each folder, and that's the "elements to generate" section.  Every other setting has to be kept in sync across all the folders; e.g., if I set a TableRename callback on the Entities .tt file, I have to copy/paste that over to the Context and Mapping .tt files as well.  Tedious and error-prone.

What I did to avoid the copy/paste was to move all of the common settings code (everything between the version # comment at the top, and the "nothing else to configure" comment at the bottom) into its own .ttinclude file, EXCEPT for the "elements to generate" section, which stayed behind in the folder-based .tt files.  I then simply included the new .ttinclude file at the top of each .tt file.  The setup looks like this now:

![poco](https://user-images.githubusercontent.com/5384706/46824227-32574480-cd5e-11e8-9945-0685e0d9c2a3.png)

Of course, this introduces a new issue, in that saving the new .ttinclude file doesn't regenerate the three folders; for that I have to open/save each of the .tt files (and a simple Rebuild doesn't trigger regeneration).  If I reverse the setup and make the common file .tt and the folder-specific files .ttinclude, then it does trigger on save, but it puts the files in the wrong place.

The enhancement would essentially allow us to have a single .tt file, and have the "elements to generate" settings include the ability to tell the tool into which folders to put the files.  Interestingly, the comment there does in fact state that it doesn't create the files in these locations, though I can't imagine that this would be such a difficult change to make.

Thanks!

**Comment by sjh37** [comment]
Similar to #104

**Comment by sjh37** [comment]
Two solutions:

1. Have a common project level `project.ttinclude` file, and several smaller .tt files which include the  `project.ttinclude` file. The smaller .tt files specify what to elements to generate.
2. Have a single `database.tt` and output the files to sub-folders. The names of the sub-folders can be specified in the `database.tt` file. If blank, it acts as normal, and creates files in the current folder.

I prefer option 2 as it's cleaner.

**Comment by sjh37** [comment]
Because of EF Core, and not having to modify the visual studio project file, option 2 is now possible.

**Comment by ClassyCircuit** [comment]
@sjh37  how to do the 2nd option in EF Core? I don't see any configuration option that allows generating into subfolders from a single .tt file.

**Comment by sjh37** [comment]
The 2nd option has not been implemented yet. When it is, it will only be for. Net core and not 4.x

**Comment by wps-preston** [comment]
Any word on Option 2 being completed for .net core???

**Comment by sjh37** [comment]
It won't take long to implement. I'll take a look today.

**Comment by wps-preston** [comment]
Thank you that would be awesome.

**Comment by wps-preston** [comment]
Were you able to get time to look at this and make the updates

**Comment by sjh37** [comment]
Hi @mbouhuys Sorry no. A more important feature came around (SQL Sequences - case 668) which caused errors. This is now fixed but not released. I need to spend one more day testing sequences, then I can come back and look into this case. Sorry for the delay, but so many good things to add.

**Comment by wps-preston** [comment]
Thank you so much for the update Simon I totally understand and I’m excited for when this feature get done…

Michael

**Comment by wps-preston** [comment]
Hello Simon, any progress on this feature???

Thank Michael

**Comment by sjh37** [comment]
Hi @mbouhuys Not yet. It's still on my todo list. Currently sorting out a Fake db context query issue...

**Comment by wps-preston** [comment]
Hello Simon,
Just pinging you to see if there has been any movement on this request?

Thanks
Michael

**Comment by sjh37** [comment]
Hi @mbouhuys Really sorry, this case is still on my todo (top of the nice to have) list. I know I must do this case as its something that a lot of developers, including myself, would like to have. It shouldn't take long to implement so I'll try and inject it between higher level cases.

**Comment by wps-preston** [comment]
Just looking for an update any progress on this???

**Comment by sjh37** [comment]
Sorry, this one kept being put back due to higher priority cases. It is fairly straight forward tbh. I'll see if I can get it done this week. I said that last time too! But I promise to see if I can have this added.

**Comment by MarkLFT** [comment]
Hi @sjh37,

Many thanks for the latest release, was wondering if you had a chance in the latest release to include the option to specify folders in the settings to allow the use of sub-folders.  At the moment I create the include and multiple tt files, but it is very messy and time consuming.

**Comment by wps-preston** [comment]
Any updates on this ??

**Comment by sjh37** [comment]
Hi @mbouhuys This will be in the next release which will be this month (December 2021).

**Comment by sjh37** [comment]
This feature has been added and will be in the next release.
If you want it now, view the code and grab the latest `EF.Reverse.POCO.v3.ttinclude` file from GitHub, and add the following to your `<database>.tt` file.

```c#
// Generate files in sub-folders ******************************************************************************************************
// Only activated if Settings.GenerateSeparateFiles = true && Settings.FileManagerType = FileManagerType.EfCore
Settings.ContextFolder           = @"";              // Sub-folder you would like your DbContext to be added to.              e.g. @"Data"
Settings.InterfaceFolder         = @"Interface";     // Sub-folder you would like your Interface to be added to.              e.g. @"Data\Interface"
Settings.PocoFolder              = @"Entities";      // Sub-folder you would like your Poco's to be added to.                 e.g. @"Data\Entities"
Settings.PocoConfigurationFolder = @"Configuration"; // Sub-folder you would like your Configuration mappings to be added to. e.g. @"Data\Configuration"
```

You can even specify multiple sub-folders if you wish and they will all be created.
```c#
Settings.PocoFolder = @"Data\Entities";
```

The generated code all has the same namespace and do not inherit the folder name as part of the namespace. This keeps the generated code in the same namespace as it was previously so nothing will break, but is now nicely organised.

**Comment by sjh37** [comment]
Now released in v3.6.0
See new AddEnum() function in the `<database>.tt` file.
